django/utils/choices.py

```diff
@@ -0,0 +1,63 @@
+from collections.abc import Callable, Iterable, Iterator, Mapping
+
+from django.db.models.enums import ChoicesMeta
+from django.utils.functional import Promise
+
+
+class ChoiceIterator:
+    """Base class for lazy iterators for choices."""
+
+
+class CallableChoiceIterator(ChoiceIterator):
+    """Iterator to lazily normalize choices generated by a callable."""
+
+    def __init__(self, choices_func):
+        self.choices_func = choices_func
+
+    def __iter__(self):
+        yield from normalize_choices(self.choices_func())
+
+
+def normalize_choices(value, *, _depth=0):
+    """Normalize choices values consistently for fields and widgets."""
+
+    match value:
```

Comment:
I guess it's perspective. This way was easier to understand compared to an earlier version using if/elif statements, but things changed a bit since then... Alas, I've run out of time for the next couple of weeks to experiment with switching this around and seeing if it's better or not 🤷🏻